Client/WarFare/UIItemUpgrade.cpp

```diff
@@ -0,0 +1,1289 @@
+	m_pUITooltipDlg->InitPos();
+	m_pUITooltipDlg->SetVisible(FALSE);
+
+	CN3UIWndBase::InitIconWnd(eWnd);
+
+	m_pStrMyGold = (CN3UIString*) GetChildByID("text_gold");
+	__ASSERT(m_pStrMyGold, "NULL UI Component!!");
+	if (m_pStrMyGold) m_pStrMyGold->SetString("0");
+}
+
+__IconItemSkill* CUIItemUpgrade::GetHighlightIconItem(CN3UIIcon* pUIIcon)
+{
+	int i;
+	for (i = 0; i < MAX_ITEM_UPGRADE_SLOT; i++)
+	{
+		if ((m_pMyUpgradeSLot[i] != nullptr) && (m_pMyUpgradeSLot[i]->pUIIcon == pUIIcon))
+			return m_pMyUpgradeSLot[i];
+	}
+
+	for (i = 0; i < MAX_ITEM_INVENTORY; i++)
+	{
+		if ((m_pMyUpgradeInv[i] != nullptr) && (m_pMyUpgradeInv[i]->pUIIcon == pUIIcon))
+			return m_pMyUpgradeInv[i];
+	}
+	if (m_pUpgradeItemSlot && m_pUpgradeItemSlot->pUIIcon == pUIIcon)
+		return m_pUpgradeItemSlot;
+
+	return nullptr;
+}
+
+void CUIItemUpgrade::Open()
+{
+	SetVisible(true);
+	ItemMoveFromInvToThis();
+
+	GoldUpdate();
+}
+
+void CUIItemUpgrade::GoldUpdate()
+{
+	if (m_pStrMyGold)
+	{
+		m_pStrMyGold->SetStringAsInt(CGameBase::s_pPlayer->m_InfoExt.iGold);
+	}
+}
+
+void CUIItemUpgrade::ItemMoveFromInvToThis()
+{
+	CUIInventory* pInven = CGameProcedure::s_pProcMain->m_pUIInventory;
+	if (!pInven)
+		return;
+
+
+	for (int i = 0; i < MAX_ITEM_INVENTORY; i++)
+	{
+		if (pInven->m_pMyInvWnd[i])
+		{
+			__IconItemSkill* spItem = pInven->m_pMyInvWnd[i];
+			spItem->pUIIcon->SetParent(this);
+
+			pInven->m_pMyInvWnd[i] = nullptr;
+			CN3UIArea* pArea;
+
+			char szID[32];
+			sprintf(szID, "a_slot_%d", i);
+			pArea = (CN3UIArea*)GetChildByID(szID);
+			if (pArea)
+			{
+				spItem->pUIIcon->SetRegion(pArea->GetRegion());
+				spItem->pUIIcon->SetMoveRect(pArea->GetRegion());
+			}
+
+			m_pMyUpgradeInv[i] = spItem;
+		}
+		// Backup the inventory state for restoration if needed.
+		m_pBackupUpgradeInv[i] = nullptr;
+		if (m_pMyUpgradeInv[i])
+		{
+			m_pBackupUpgradeInv[i] = new __IconItemSkill(*m_pMyUpgradeInv[i]);
+		}
+	}
+}
+
+void CUIItemUpgrade::Close()
+{
+	if (IsVisible())
+		SetVisible(false);
+	RestoreInventoryFromBackup();
+	AnimClose();
+
+	if (GetState() == UI_STATE_ICON_MOVING)
+		IconRestore();
+	SetState(UI_STATE_COMMON_NONE);
+	CN3UIWndBase::AllHighLightIconFree();
+
+	ItemMoveFromThisToInv();
+}
+
+void CUIItemUpgrade::ItemMoveFromThisToInv()
+{
+	CUIInventory* pInven = CGameProcedure::s_pProcMain->m_pUIInventory;
+	if (!pInven)
+		return;
+
+	for (int i = 0; i < MAX_ITEM_INVENTORY; i++)
+	{
+		if (m_pMyUpgradeInv[i])
+		{
+			__IconItemSkill* spItem = m_pMyUpgradeInv[i];
+			spItem->pUIIcon->SetParent(pInven);
+
+			m_pMyUpgradeInv[i] = nullptr;
+
+			CN3UIArea* pArea;
+
+			pArea = pInven->GetChildAreaByiOrder(UI_AREA_TYPE_INV, i);
+			if (pArea)
+			{
+				spItem->pUIIcon->SetRegion(pArea->GetRegion());
+				spItem->pUIIcon->SetMoveRect(pArea->GetRegion());
+			}
+			pInven->m_pMyInvWnd[i] = spItem;
+		}
+	}
+}
+
+bool CUIItemUpgrade::ReceiveIconDrop(__IconItemSkill* spItem, POINT ptCur)
+{
+	// Temp Define 
+#define FAIL_RETURN {	\
```

Comment:
This macro is bad and should not be perpetuated.

It serves no real purpose; it returns false, only for the caller to call `CancelIconDrop()`, which calls `AllHighLightIconFree()` and `SetState(UI_STATE_COMMON_NONE)` anyway.

So just remove this macro and call return false directly. It does not need to exist.